// vue-cesium/es/components/providers/baidu/index.mjs
import { defineComponent, getCurrentInstance, createCommentVNode } from 'vue';
import BaiduMapImageryProvider from './BaiduMapImageryProvider.mjs';
import '../../../composables/index.mjs';
import { url, rectangle, ellipsoid, tileDiscardPolicy, credit, minimumLevel, maximumLevel } from '../../../utils/cesium-props.mjs';
import { kebabCase } from '../../../utils/util.mjs';
import { providerEmits } from '../../../utils/emits.mjs';
import useProviders from '../../../composables/use-providers/index.mjs';

"use strict";
const baiduImageryProviderProps = {
  ...url,
  ...rectangle,
  ...ellipsoid,
  ...tileDiscardPolicy,
  ...credit,
  ...minimumLevel,
  ...maximumLevel,
  projectionTransforms: {
    type: [Boolean, Object],
    default: () => {
      return {
        from: "BD09",
        to: "WGS84"
      };
    }
  },
  scale: {
    type: String,
    default: "2"
  },
  ak: {
    type: String,
    default: "5ieMMexWmzB9jivTq6oCRX9j"
  },
  subdomains: {
    type: Array,
    default: () => ["0", "1", "2", "3"]
  },
  // https://lbsyun.baidu.com/custom/list.htm
  mapStyle: {
    type: String,
    default: "vec"
    // img vec traffic normal light dark redalert googlelite grassgreen midnight pink darkgreen bluish grayscale hardedge
  },
  qt: {
    type: String,
    default: "vtile"
  },
  styles: {
    type: String,
    // sl 背景透明 pl 正常 ph 大字体
    default: "pl"
  },
  showtext: {
    type: String,
    // 0 不显示， 1 显示
    default: "1"
  }
};
var ImageryProviderBaidu = defineComponent({
  name: "VcImageryProviderBaidu",
  props: baiduImageryProviderProps,
  emits: providerEmits,
  setup(props, ctx) {
    const instance = getCurrentInstance();
    instance.cesiumClass = "BaiduMapImageryProvider";
    const providersState = useProviders(props, ctx, instance);
    if (void 0 === providersState) {
      return;
    }
    instance.createCesiumObject = async () => {
      Cesium.BaiduMapImageryProvider = Cesium.BaiduMapImageryProvider || BaiduMapImageryProvider;
      if (providersState.unwatchFns.length === 0) {
        providersState.setPropsWatcher(true);
      }
      const options = providersState.transformProps(props);
      return new Cesium.BaiduMapImageryProvider(options);
    };
    return () => {
      var _a;
      return createCommentVNode(kebabCase(((_a = instance.proxy) == null ? void 0 : _a.$options.name) || ""));
    };
  }
});

export { baiduImageryProviderProps, ImageryProviderBaidu as default };
