// vue-cesium/lib/components/primitives/osm-buildings/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
require('../../../composables/index.js');
var cesiumProps = require('../../../utils/cesium-props.js');
var util = require('../../../utils/util.js');
var emits = require('../../../utils/emits.js');
var cesiumHelpers = require('../../../utils/cesium-helpers.js');
var index = require('../../../composables/use-primitives/index.js');

"use strict";
const osmBuildingsProps = {
  ...cesiumProps.defaultColor,
  ...cesiumProps.tileStyle,
  enableShowOutline: {
    type: Boolean,
    default: true
  },
  showOutline: {
    type: Boolean,
    default: true
  },
  ...cesiumProps.enableMouseEvent
};
var PrimitiveOsmBuildings = vue.defineComponent({
  name: "VcPrimitiveOsmBuildings",
  props: osmBuildingsProps,
  emits: {
    ...emits.primitiveEmits
  },
  setup(props, ctx) {
    const instance = vue.getCurrentInstance();
    instance.cesiumClass = "VcPrimitiveOsmBuildings";
    const primitivesState = index["default"](props, ctx, instance);
    instance.createCesiumObject = async () => {
      const options = primitivesState == null ? void 0 : primitivesState.transformProps(props);
      options.style = options.tileStyle;
      delete options.tileStyle;
      return cesiumHelpers.compareCesiumVersion(Cesium.VERSION, "1.104") ? await Cesium.createOsmBuildingsAsync(options) : Cesium.createOsmBuildings(options);
    };
    return () => {
      var _a;
      return vue.createCommentVNode(util.kebabCase(((_a = instance.proxy) == null ? void 0 : _a.$options.name) || ""));
    };
  }
});

exports["default"] = PrimitiveOsmBuildings;
exports.osmBuildingsProps = osmBuildingsProps;
