// vue-cesium/es/components/providers/ion/index.mjs
import { defineComponent, getCurrentInstance, createCommentVNode } from 'vue';
import '../../../composables/index.mjs';
import { accessToken } from '../../../utils/cesium-props.mjs';
import { kebabCase } from '../../../utils/util.mjs';
import { providerEmits } from '../../../utils/emits.mjs';
import { compareCesiumVersion } from '../../../utils/cesium-helpers.mjs';
import useProviders from '../../../composables/use-providers/index.mjs';

"use strict";
const ionImageryProviderProps = {};
var ImageryProviderIon = defineComponent({
  name: "VcImageryProviderIon",
  props: {
    assetId: Number,
    ...accessToken,
    server: [String, Object]
  },
  emits: providerEmits,
  setup(props, ctx) {
    const instance = getCurrentInstance();
    instance.cesiumClass = "IonImageryProvider";
    const providersState = useProviders(props, ctx, instance);
    if (void 0 === providersState) {
      return;
    }
    instance.createCesiumObject = async () => {
      const options = providersState.transformProps(props);
      if (compareCesiumVersion(Cesium.VERSION, "1.104") && typeof Cesium[instance.cesiumClass].fromAssetId === "function") {
        return await Cesium.IonImageryProvider.fromAssetId(options.assetId, options);
      } else {
        return new Cesium.IonImageryProvider(options);
      }
    };
    return () => {
      var _a;
      return createCommentVNode(kebabCase(((_a = instance.proxy) == null ? void 0 : _a.$options.name) || ""));
    };
  }
});

export { ImageryProviderIon as default, ionImageryProviderProps };
