// vue-cesium/es/components/primitives/i3s/index.mjs
import { defineComponent, getCurrentInstance, createCommentVNode } from 'vue';
import '../../../composables/index.mjs';
import { url, show, enableMouseEvent } from '../../../utils/cesium-props.mjs';
import { kebabCase } from '../../../utils/util.mjs';
import { primitiveEmits } from '../../../utils/emits.mjs';
import usePrimitives from '../../../composables/use-primitives/index.mjs';

"use strict";
const i3sDataProviderProps = {
  ...url,
  name: String,
  ...show,
  geoidTiledTerrainProvider: {
    type: Object
  },
  traceFetches: {
    type: Boolean,
    default: false
  },
  cesium3dTilesetOptions: {
    type: Object
  },
  ...enableMouseEvent
};
var PrimitiveI3sDataProvider = defineComponent({
  name: "VcPrimitiveI3sDataProvider",
  props: i3sDataProviderProps,
  emits: primitiveEmits,
  setup(props, ctx) {
    const instance = getCurrentInstance();
    instance.cesiumClass = "I3SDataProvider";
    usePrimitives(props, ctx, instance);
    return () => {
      var _a;
      return createCommentVNode(kebabCase(((_a = instance.proxy) == null ? void 0 : _a.$options.name) || ""));
    };
  }
});

export { PrimitiveI3sDataProvider as default, i3sDataProviderProps };
