// vue-cesium/lib/components/providers/ion/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
require('../../../composables/index.js');
var cesiumProps = require('../../../utils/cesium-props.js');
var util = require('../../../utils/util.js');
var emits = require('../../../utils/emits.js');
var cesiumHelpers = require('../../../utils/cesium-helpers.js');
var index = require('../../../composables/use-providers/index.js');

"use strict";
const ionImageryProviderProps = {};
var ImageryProviderIon = vue.defineComponent({
  name: "VcImageryProviderIon",
  props: {
    assetId: Number,
    ...cesiumProps.accessToken,
    server: [String, Object]
  },
  emits: emits.providerEmits,
  setup(props, ctx) {
    const instance = vue.getCurrentInstance();
    instance.cesiumClass = "IonImageryProvider";
    const providersState = index["default"](props, ctx, instance);
    if (void 0 === providersState) {
      return;
    }
    instance.createCesiumObject = async () => {
      const options = providersState.transformProps(props);
      if (cesiumHelpers.compareCesiumVersion(Cesium.VERSION, "1.104") && typeof Cesium[instance.cesiumClass].fromAssetId === "function") {
        return await Cesium.IonImageryProvider.fromAssetId(options.assetId, options);
      } else {
        return new Cesium.IonImageryProvider(options);
      }
    };
    return () => {
      var _a;
      return vue.createCommentVNode(util.kebabCase(((_a = instance.proxy) == null ? void 0 : _a.$options.name) || ""));
    };
  }
});

exports["default"] = ImageryProviderIon;
exports.ionImageryProviderProps = ionImageryProviderProps;
