// vue-cesium/lib/components/providers/amap/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
require('../../../composables/index.js');
var cesiumProps = require('../../../utils/cesium-props.js');
var util = require('../../../utils/util.js');
var emits = require('../../../utils/emits.js');
var AMapImageryProvider = require('./AMapImageryProvider.js');
var index = require('../../../composables/use-providers/index.js');

"use strict";
const amapImageryProviderProps = {
  url: {
    type: String,
    default: "https://{domain}{s}.is.autonavi.com/appmaptile?lang={lang}&size=1&style={style}&scl={scl}&ltype={ltype}&x={x}&y={y}&z={z}"
  },
  subdomains: {
    type: Array,
    default: () => ["01", "02", "03", "04"]
  },
  domain: {
    type: String,
    default: "webst"
  },
  lang: {
    type: String,
    default: "zh_cn"
  },
  mapStyle: {
    // 地图类型控制，6卫星图（st），7简图（st rd），8详图（不透明rd，透明图st）
    type: String,
    default: "6"
  },
  scl: {
    // 尺寸控制，1=256,2=512
    type: String,
    default: "1"
  },
  ltype: {
    // 线性控制，只对地图要素进行控制，没有文字注记，要素多少，是否透明
    // 纯道路 ltype=11 mapStyle=8
    // 纯地标  ltype=4 mapStyle=8
    // 道路标注 ltype=0 mapStyle=8
    type: String,
    default: "0"
  },
  ...cesiumProps.credit,
  ...cesiumProps.minimumLevel,
  ...cesiumProps.maximumLevel,
  ...cesiumProps.rectangle,
  ...cesiumProps.tilingScheme,
  ...cesiumProps.projectionTransforms
};
var ImageryProviderAmap = vue.defineComponent({
  name: "VcImageryProviderAmap",
  props: amapImageryProviderProps,
  emits: emits.providerEmits,
  setup(props, ctx) {
    const instance = vue.getCurrentInstance();
    instance.cesiumClass = "AMapImageryProvider";
    const providersState = index["default"](props, ctx, instance);
    if (void 0 === providersState) {
      return;
    }
    instance.createCesiumObject = async () => {
      Cesium.AMapImageryProvider = Cesium.AMapImageryProvider || AMapImageryProvider["default"];
      if (providersState.unwatchFns.length === 0) {
        providersState.setPropsWatcher(true);
      }
      const options = providersState.transformProps(props);
      return new Cesium.AMapImageryProvider(options);
    };
    return () => {
      var _a;
      return vue.createCommentVNode(util.kebabCase(((_a = instance.proxy) == null ? void 0 : _a.$options.name) || ""));
    };
  }
});

exports.amapImageryProviderProps = amapImageryProviderProps;
exports["default"] = ImageryProviderAmap;
