// vue-cesium/lib/components/primitives/voxel/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
require('../../../composables/index.js');
var cesiumProps = require('../../../utils/cesium-props.js');
var util = require('../../../utils/util.js');
var emits = require('../../../utils/emits.js');
var index = require('../../../composables/use-primitives/index.js');

"use strict";
const voxelPromitiveProps = {
  provider: {
    type: Object
  },
  ...cesiumProps.modelMatrix,
  ...cesiumProps.customShader,
  ...cesiumProps.clock,
  ...cesiumProps.enableMouseEvent
};
var PrimitiveVoxel = vue.defineComponent({
  name: "VcPrimitiveVoxel",
  props: voxelPromitiveProps,
  emits: emits.primitiveEmits,
  setup(props, ctx) {
    const instance = vue.getCurrentInstance();
    instance.cesiumClass = "VoxelPrimitive";
    index["default"](props, ctx, instance);
    return () => {
      var _a;
      return vue.createCommentVNode(util.kebabCase(((_a = instance.proxy) == null ? void 0 : _a.$options.name) || ""));
    };
  }
});

exports["default"] = PrimitiveVoxel;
exports.voxelPromitiveProps = voxelPromitiveProps;
