// vue-cesium/es/components/primitives/index.mjs
import PrimitiveClassification from './classification/index.mjs';
export { classificationPrimitiveProps } from './classification/index.mjs';
import PrimitiveGround from './ground/index.mjs';
export { groundPrimitiveProps } from './ground/index.mjs';
import PrimitiveGroundPolyline from './ground-polyline/index.mjs';
export { groundPolylinePrimitiveProps } from './ground-polyline/index.mjs';
import PrimitiveModel from './model/index.mjs';
export { modelPrimitiveProps } from './model/index.mjs';
import Primitive from './primitive/index.mjs';
export { primitiveProps } from './primitive/index.mjs';
import PrimitiveTileset from './tileset/index.mjs';
export { tilesetPrimitiveProps } from './tileset/index.mjs';
import PrimitiveOsmBuildings from './osm-buildings/index.mjs';
export { osmBuildingsProps } from './osm-buildings/index.mjs';
import PrimitiveParticle from './particle/index.mjs';
export { particlePrimitiveProps } from './particle/index.mjs';
import PrimitiveViewshed from './viewshed/index.mjs';
export { viewshedProps } from './viewshed/index.mjs';
import PrimitiveTimeDynamicPointCloud from './time-dynamic-point-cloud/index.mjs';
export { timeDynamicPointCloudProps } from './time-dynamic-point-cloud/index.mjs';
import PrimitiveI3sDataProvider from './i3s/index.mjs';
export { i3sDataProviderProps } from './i3s/index.mjs';
import PrimitiveVoxel from './voxel/index.mjs';
export { voxelPromitiveProps } from './voxel/index.mjs';
import PrimitiveCluster from './primitive-cluster/index.mjs';
export { primitiveClusterProps } from './primitive-cluster/index.mjs';

"use strict";
const components = [
  PrimitiveClassification,
  PrimitiveGround,
  PrimitiveGroundPolyline,
  PrimitiveModel,
  Primitive,
  PrimitiveTileset,
  PrimitiveOsmBuildings,
  PrimitiveI3sDataProvider,
  PrimitiveVoxel,
  PrimitiveTimeDynamicPointCloud,
  PrimitiveParticle,
  PrimitiveCluster
];
const install = (app) => {
  components.forEach((cmp) => {
    app.component(cmp.name, cmp);
  });
};
var index = {
  install
};
components.forEach((cmp) => {
  cmp["install"] = (app) => {
    app.component(cmp.name, cmp);
  };
});
const VcPrimitiveClassification = PrimitiveClassification;
const VcPrimitiveGround = PrimitiveGround;
const VcPrimitiveGroundPolyline = PrimitiveGroundPolyline;
const VcPrimitiveModel = PrimitiveModel;
const VcPrimitive = Primitive;
const VcPrimitiveTileset = PrimitiveTileset;
const VcPrimitiveOsmBuildings = PrimitiveOsmBuildings;
const VcPrimitiveParticle = PrimitiveParticle;
const VcViewshed = PrimitiveViewshed;
const VcPrimitiveTimeDynamicPointCloud = PrimitiveTimeDynamicPointCloud;
const VcPrimitiveI3sDataProvider = PrimitiveI3sDataProvider;
const VcPrimitiveVoxel = PrimitiveVoxel;
const VcPrimitiveCluster = PrimitiveCluster;

export { VcPrimitive, VcPrimitiveClassification, VcPrimitiveCluster, VcPrimitiveGround, VcPrimitiveGroundPolyline, VcPrimitiveI3sDataProvider, VcPrimitiveModel, VcPrimitiveOsmBuildings, VcPrimitiveParticle, VcPrimitiveTileset, VcPrimitiveTimeDynamicPointCloud, VcPrimitiveVoxel, VcViewshed, index as default };
