// vue-cesium/lib/components/primitives/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var index$1 = require('./classification/index.js');
var index$2 = require('./ground/index.js');
var index$3 = require('./ground-polyline/index.js');
var index$4 = require('./model/index.js');
var index$7 = require('./primitive/index.js');
var index$8 = require('./tileset/index.js');
var index$5 = require('./osm-buildings/index.js');
var index$6 = require('./particle/index.js');
var index$9 = require('./viewshed/index.js');
var index$a = require('./time-dynamic-point-cloud/index.js');
var index$b = require('./i3s/index.js');
var index$c = require('./voxel/index.js');
var index$d = require('./primitive-cluster/index.js');

"use strict";
const components = [
  index$1["default"],
  index$2["default"],
  index$3["default"],
  index$4["default"],
  index$7["default"],
  index$8["default"],
  index$5["default"],
  index$b["default"],
  index$c["default"],
  index$a["default"],
  index$6["default"],
  index$d["default"]
];
const install = (app) => {
  components.forEach((cmp) => {
    app.component(cmp.name, cmp);
  });
};
var index = {
  install
};
components.forEach((cmp) => {
  cmp["install"] = (app) => {
    app.component(cmp.name, cmp);
  };
});
const VcPrimitiveClassification = index$1["default"];
const VcPrimitiveGround = index$2["default"];
const VcPrimitiveGroundPolyline = index$3["default"];
const VcPrimitiveModel = index$4["default"];
const VcPrimitive = index$7["default"];
const VcPrimitiveTileset = index$8["default"];
const VcPrimitiveOsmBuildings = index$5["default"];
const VcPrimitiveParticle = index$6["default"];
const VcViewshed = index$9["default"];
const VcPrimitiveTimeDynamicPointCloud = index$a["default"];
const VcPrimitiveI3sDataProvider = index$b["default"];
const VcPrimitiveVoxel = index$c["default"];
const VcPrimitiveCluster = index$d["default"];

exports.classificationPrimitiveProps = index$1.classificationPrimitiveProps;
exports.groundPrimitiveProps = index$2.groundPrimitiveProps;
exports.groundPolylinePrimitiveProps = index$3.groundPolylinePrimitiveProps;
exports.modelPrimitiveProps = index$4.modelPrimitiveProps;
exports.primitiveProps = index$7.primitiveProps;
exports.tilesetPrimitiveProps = index$8.tilesetPrimitiveProps;
exports.osmBuildingsProps = index$5.osmBuildingsProps;
exports.particlePrimitiveProps = index$6.particlePrimitiveProps;
exports.viewshedProps = index$9.viewshedProps;
exports.timeDynamicPointCloudProps = index$a.timeDynamicPointCloudProps;
exports.i3sDataProviderProps = index$b.i3sDataProviderProps;
exports.voxelPromitiveProps = index$c.voxelPromitiveProps;
exports.primitiveClusterProps = index$d.primitiveClusterProps;
exports.VcPrimitive = VcPrimitive;
exports.VcPrimitiveClassification = VcPrimitiveClassification;
exports.VcPrimitiveCluster = VcPrimitiveCluster;
exports.VcPrimitiveGround = VcPrimitiveGround;
exports.VcPrimitiveGroundPolyline = VcPrimitiveGroundPolyline;
exports.VcPrimitiveI3sDataProvider = VcPrimitiveI3sDataProvider;
exports.VcPrimitiveModel = VcPrimitiveModel;
exports.VcPrimitiveOsmBuildings = VcPrimitiveOsmBuildings;
exports.VcPrimitiveParticle = VcPrimitiveParticle;
exports.VcPrimitiveTileset = VcPrimitiveTileset;
exports.VcPrimitiveTimeDynamicPointCloud = VcPrimitiveTimeDynamicPointCloud;
exports.VcPrimitiveVoxel = VcPrimitiveVoxel;
exports.VcViewshed = VcViewshed;
exports["default"] = index;
