// vue-cesium/lib/components/primitives/i3s/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
require('../../../composables/index.js');
var cesiumProps = require('../../../utils/cesium-props.js');
var util = require('../../../utils/util.js');
var emits = require('../../../utils/emits.js');
var index = require('../../../composables/use-primitives/index.js');

"use strict";
const i3sDataProviderProps = {
  ...cesiumProps.url,
  name: String,
  ...cesiumProps.show,
  geoidTiledTerrainProvider: {
    type: Object
  },
  traceFetches: {
    type: Boolean,
    default: false
  },
  cesium3dTilesetOptions: {
    type: Object
  },
  ...cesiumProps.enableMouseEvent
};
var PrimitiveI3sDataProvider = vue.defineComponent({
  name: "VcPrimitiveI3sDataProvider",
  props: i3sDataProviderProps,
  emits: emits.primitiveEmits,
  setup(props, ctx) {
    const instance = vue.getCurrentInstance();
    instance.cesiumClass = "I3SDataProvider";
    index["default"](props, ctx, instance);
    return () => {
      var _a;
      return vue.createCommentVNode(util.kebabCase(((_a = instance.proxy) == null ? void 0 : _a.$options.name) || ""));
    };
  }
});

exports["default"] = PrimitiveI3sDataProvider;
exports.i3sDataProviderProps = i3sDataProviderProps;
